src/main/java/frc/robot/subsystems/ShooterSubsystem.java

```diff
@@ -0,0 +1,181 @@
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine;
+import edu.wpi.first.wpilibj2.command.sysid.SysIdRoutine.Direction;
+import frc.robot.Constants.IndexerConstants;
+import frc.robot.subsystems.sysid.SysIdRoutineSignalLogger;
+
+public class ShooterSubsystem extends SubsystemBase {
+  private final TalonFX shooterLeftMotor = new TalonFX(DEVICE_ID_SHOOTER_LEFT, CANIVORE_BUS_NAME);
+  private final TalonFX shooterRightMotor = new TalonFX(DEVICE_ID_SHOOTER_RIGHT, CANIVORE_BUS_NAME);
+  private final CANSparkMax actuatorMotor = new CANSparkMax(DEVICE_ID_ACTUATOR_MOTOR, MotorType.kBrushless);
+  // creates new devices and sets them to the device Ids set in
+  // Constants.java
+
+  private static final double ENCODER_OFFSET = -0.2285f;
+  private static final float LIMIT_BOTTOM = 0.12f;
+  private static final float LIMIT_TOP = 0.4272f;
+  private static final double LIMIT_TOP_RADIANS = Units.rotationsToRadians(LIMIT_TOP + ENCODER_OFFSET);
+  private static final double LIMIT_BOTTOM_RADIANS = Units.rotationsToRadians(LIMIT_BOTTOM + ENCODER_OFFSET);
+  private boolean isActiveStopped = false;
+  public boolean hasRing = false;
+  private SparkAbsoluteEncoder actuatorEncoder;
+  private final SparkPIDController pidController;
+  private VoltageOut voltageRequest = new VoltageOut(0).withEnableFOC(true);
+
+  private SysIdRoutine shooterMotorSysIdRoutine = new SysIdRoutine(
+      new SysIdRoutine.Config(null, null, null, SysIdRoutineSignalLogger.logState()),
+      new SysIdRoutine.Mechanism((volts) -> {
+        shooterRightMotor.setControl(voltageRequest.withOutput(volts.in(Volts)));
+        shooterLeftMotor.setControl(voltageRequest.withOutput(volts.in(Volts)));
+      }, null, this));
+
+  private final VelocityTorqueCurrentFOC shooterMotorVelocity = new VelocityTorqueCurrentFOC(0, 0, 0, 1, false, false,
+      false);
+
+  public ShooterSubsystem() {
+    var shooterMotorConfig = new TalonFXConfiguration();
+    shooterMotorConfig.Slot0 = Slot0Configs.from(SHOOTER_SLOT_CONFIGS);
+    shooterMotorConfig.MotorOutput.NeutralMode = Coast;
+    shooterMotorConfig.OpenLoopRamps.DutyCycleOpenLoopRampPeriod = 1;
+    shooterMotorConfig.MotorOutput.Inverted = InvertedValue.Clockwise_Positive;
+    shooterLeftMotor.getConfigurator().apply(shooterMotorConfig);
+    shooterMotorConfig.MotorOutput.Inverted = InvertedValue.CounterClockwise_Positive;
+    shooterRightMotor.getConfigurator().apply(shooterMotorConfig);
+
+    // Get the through-bore-encoder absolute encoder
+    actuatorEncoder = actuatorMotor.getAbsoluteEncoder(kDutyCycle);
+    actuatorEncoder.setInverted(true);
+    actuatorEncoder.setAverageDepth(64);
+    pidController = actuatorMotor.getPIDController();
+    pidController.setFeedbackDevice(actuatorEncoder);
+
+  }
+
+  private static SparkPIDController indexerMotorConfig(CANSparkMax sparkMax, boolean invert) {

```

Comment:
This looks like it was copied from the indexer. It should be reviewed, and a separate method isn't really needed since there is only one Spark Max controller.